### ISSUE TITLE
failed to compile the directive "secrule": invalid action "proxy"

### ISSUE THREAD

**Issue by titpetric** [corazawaf/coraza#775]
## Description

WAF config from string reports:

```
invalid WAF config from string: failed to compile the directive "secrule": invalid action "proxy"
```

Running 3.0.0-rc1.

### Steps to reproduce

https://coraza.io/docs/seclang/actions/#proxy

The rules for `block` and `redirect` are parsed and enforced correctly. Proxy is unrecognized.

### Expected result

I should get a filled out types.Interruption with Action: proxy and Data: url;

### Actual result

The above error.

**Comment by jptosso** [comment]
Hey @titpetric , thank you for your report.

The action proxy is not implemented because it highly relies on the connector. In the meantime you can easily implement it with the following code:

```go
// Copyright 2022 Juan Pablo Tosso and the OWASP Coraza contributors
// SPDX-License-Identifier: Apache-2.0

package actions

import (
	"github.com/corazawaf/coraza/v3/rules"
	"github.com/corazawaf/coraza/v3/types"
	"github.com/corazawaf/coraza/v3/experimental/plugins"
)

type denyFn struct{
      target string
}

func (a *proxyFn) Init(_ rules.RuleMetadata, data string) error {
	if len(data) == 0 {
		return ErrUnexpectedArguments
	}
        a.target = data
	return nil
}

func (a *proxyFn) Evaluate(r rules.RuleMetadata, tx rules.TransactionState) {
	rid := r.ID()
	if rid == 0 {
		rid = r.ParentID()
	}
	tx.Interrupt(&types.Interruption{
		Status: r.Status(),
		RuleID: rid,
		Action: "deny",
                Data: a.target,
	})
}

func (a *proxyFn) Type() rules.ActionType {
	return rules.ActionTypeDisruptive
}

func proxy() rules.Action {
	return &proxyFn{}
}

var (
	_ rules.Action      = &proxyFn{}
	_ ruleActionWrapper = proxy
)

func init() {
      plugins.RegisterAction("deny", proxy)
}
```

@jcchavezs @fzipi any thoughts, should we add this code to the core?

**Comment by titpetric** [comment]
I don't really need it, but it's documented, and maybe I would have used it if it worked. If it gets fixed for 3.0.0 that works for me, happy to look at it and contribute a PR if you like. I did verify `redirect` works, which is essentially the same deal.

**Comment by jcchavezs** [comment]
Hi @titpetric thanks for coming by! I wonder if this proxy action should be part of coraza and or if it should be part of a WAF itself. In the past we dropped certain functionality that was handy (in the Apache days) but it looked like out of the scope of a WAF library, for example content injection. To me this seems out of the scope of a WAF and you can do it in many ways these days.

Any input @fzipi @dune73 @airween  @maxlemke?

**Comment by jptosso** [comment]
Based on that criteria, we should also remove `redirect` or keep both.

**Comment by jcchavezs** [comment]
Yeah probably that one too, let's see what people say. I guess it is not part of the CRS cc @M4tteoP

**Comment by theseion** [comment]
We don't use `proxy` anywhere in CRS (obviously, since we don't presume to know what should be proxied :) ). I see `proxy` and `redirect` mostly as convenience functions. I see three scenarios:
- you don't want to modify the web server configuration
- you can create new rules but not modify the web server configuration (e.g. with cloud providers; not actually sure that any of them would support these actions)
- you want to proxy / redirect based on some kind of scoring, e.g. proxy all requests with anomaly score > 5

However, especially proxy configurations can be so complex, I'm pretty sure that only the most basic ones would work with the `proxy` action. Personally, I'd leave those things to the web server.

Since the actions are disruptive, I don't think they would work in conjunction with the usual scoring + blocking logic. Once you redirect / proxy, the request will probably never reach the rule that blocks based on anomaly score. This is just guesswork though. Maybe @dune73 knows that better.

**Comment by titpetric** [comment]
Action gets evaluated at the same level as redirect for me, and i can skip
to a preexisting proxy middleware from the .Data as config at this point,
and yeah, it really just depends on the use case of what specific rules
people can come up with. Having a plugin or plugins to enable for redirect
and proxy would get me to the expected state, which is to add a few unit
tests for rules that I can expect to explain to people.

**Comment by dune73** [comment]
I do not think the proxy action is in wide use.

Adding to @theseion's use cases, the one I think the original author had in mind and the one I think we cover in the handbook is the idea to proxy certain requests to a sandbox / honeypot.

So you could integrate this nicely with CRS but instead of blocking you forward to a sandboxed system and observe the attackers there.

I have not seen this in practice, but I think it's a worthwhile setup. Yet if you can implement this easily then there is nothing wrong with removing it from the mainline in my eyes.

**Comment by jcchavezs** [comment]
We will probably remove that from documentation once we merge https://github.com/corazawaf/coraza.io/pull/167/files and do the same autogeneration with actions. Let's keep the issue as a reminder.
